File: powermanager/src/main/java/com/pyamsoft/powermanager/logger/LoggerPreferenceFragment.java
/*
 * Copyright 2017 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.powermanager.logger;

import android.os.Bundle;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.preference.Preference;
import android.support.v7.preference.PreferenceFragmentCompat;
import android.view.View;
import com.pyamsoft.powermanager.Injector;
import com.pyamsoft.powermanager.PowerManager;
import com.pyamsoft.powermanager.R;
import com.pyamsoft.powermanager.base.logger.LoggerPresenter;
import javax.inject.Inject;
import javax.inject.Named;

public class LoggerPreferenceFragment extends PreferenceFragmentCompat
    implements LoggerPresenter.DeleteCallback, LoggerPresenter.LogCallback {

  @NonNull public static final String TAG = "LoggerPreferenceFragment";
  @Inject @Named("logger_presenter_wifi") LoggerPresenter loggerWifi;
  @Inject @Named("logger_presenter_data") LoggerPresenter loggerData;
  @Inject @Named("logger_presenter_bluetooth") LoggerPresenter loggerBluetooth;
  @Inject @Named("logger_presenter_sync") LoggerPresenter loggerSync;
  @Inject @Named("logger_presenter_airplane") LoggerPresenter loggerAirplane;
  @Inject @Named("logger_presenter_doze") LoggerPresenter loggerDoze;
  @Inject @Named("logger_presenter_trigger") LoggerPresenter loggerTrigger;
  @Inject @Named("logger_presenter_manager") LoggerPresenter loggerManager;
  @Nullable private Preference loggingEnabled;

  @Override public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
    addPreferencesFromResource(R.xml.logger);
  }

  @Override public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    Injector.get().provideComponent().inject(this);
  }

  @Override public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);
    loggingEnabled = findPreference(getString(R.string.logger_enabled));
    if (loggingEnabled == null) {
      throw new NullPointerException("LoggingEnabled view is NULL");
    }

    loggingEnabled.setOnPreferenceChangeListener((preference, newValue) -> {
      if (newValue instanceof Boolean) {
        deleteAllPreviousLogs();
        return true;
      }
      return false;
    });
  }

  @SuppressWarnings("WeakerAccess") void deleteAllPreviousLogs() {
    loggerWifi.deleteLog(this);
    loggerData.deleteLog(this);
    loggerBluetooth.deleteLog(this);
    loggerSync.deleteLog(this);
    loggerAirplane.deleteLog(this);
    loggerDoze.deleteLog(this);
    loggerManager.deleteLog(this);
    loggerTrigger.deleteLog(this);
  }

  @Override public void onDestroyView() {
    super.onDestroyView();
    if (loggingEnabled != null) {
      loggingEnabled.setOnPreferenceChangeListener(null);
    }
  }

  @Override public void onStart() {
    super.onStart();
    loggerWifi.retrieveLogContents(this);
    loggerData.retrieveLogContents(this);
    loggerBluetooth.retrieveLogContents(this);
    loggerSync.retrieveLogContents(this);
    loggerAirplane.retrieveLogContents(this);
    loggerDoze.retrieveLogContents(this);
    loggerManager.retrieveLogContents(this);
    loggerTrigger.retrieveLogContents(this);
  }

  @Override public void onStop() {
    super.onStop();
    loggerWifi.stop();
    loggerData.stop();
    loggerBluetooth.stop();
    loggerSync.stop();
    loggerAirplane.stop();
    loggerDoze.stop();
    loggerManager.stop();
    loggerTrigger.stop();
  }

  @Override public void onDestroy() {
    super.onDestroy();
    loggerWifi.destroy();
    loggerData.destroy();
    loggerBluetooth.destroy();
    loggerSync.destroy();
    loggerAirplane.destroy();
    loggerDoze.destroy();
    loggerManager.destroy();
    loggerTrigger.destroy();
    PowerManager.getRefWatcher(this).watch(this);
  }

  @CheckResult @NonNull private LoggerDialog getLoggerDialog() {
    final Fragment loggerDialog = getParentFragment();
    if (loggerDialog instanceof LoggerDialog) {
      return (LoggerDialog) loggerDialog;
    } else {
      throw new RuntimeException("Parent is not LoggerDialog");
    }
  }

  @Override public void onPrepareLogContentRetrieval() {
    getLoggerDialog().onPrepareLogContentRetrieval();
  }

  @Override public void onLogContentRetrieved(@NonNull String logLine) {
    getLoggerDialog().onLogContentRetrieved(logLine);
  }

  @Override public void onAllLogContentsRetrieved() {
    getLoggerDialog().onAllLogContentsRetrieved();
  }

  @Override public void onLogDeleted(@NonNull String logId) {
    getLoggerDialog().onLogDeleted(logId);
  }
}

File: powermanager/src/main/java/com/pyamsoft/powermanager/logger/LoggerPreferenceFragment.kt
/*
 * Copyright 2017 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.powermanager.logger

import android.os.Bundle
import android.support.annotation.CheckResult
import android.support.v7.preference.Preference
import android.view.View
import com.pyamsoft.powermanager.Injector
import com.pyamsoft.powermanager.R
import com.pyamsoft.powermanager.base.logger.LoggerPreferencePresenter
import com.pyamsoft.powermanager.base.logger.LoggerPresenter
import com.pyamsoft.powermanager.uicore.WatchedPreferenceFragment
import javax.inject.Inject
import javax.inject.Named

class LoggerPreferenceFragment : WatchedPreferenceFragment() {
  @field:[Inject Named("logger_presenter_wifi")] lateinit internal var loggerWifi: LoggerPresenter
  @field:[Inject Named("logger_presenter_data")] lateinit internal var loggerData: LoggerPresenter
  @field:[Inject Named(
      "logger_presenter_bluetooth")] lateinit internal var loggerBluetooth: LoggerPresenter
  @field:[Inject Named("logger_presenter_sync")] lateinit internal var loggerSync: LoggerPresenter
  @field:[Inject Named(
      "logger_presenter_airplane")] lateinit internal var loggerAirplane: LoggerPresenter
  @field:[Inject Named("logger_presenter_doze")] lateinit internal var loggerDoze: LoggerPresenter
  @field:[Inject Named(
      "logger_presenter_trigger")] lateinit internal var loggerTrigger: LoggerPresenter
  @field:[Inject Named(
      "logger_presenter_manager")] lateinit internal var loggerManager: LoggerPresenter
  private lateinit var loggingEnabled: Preference
  @Inject internal lateinit var loggerPrefrencePresenter: LoggerPreferencePresenter

  override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
    addPreferencesFromResource(R.xml.logger)
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    Injector.with(context) {
      it.inject(this)
    }
  }

  override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    loggingEnabled = findPreference(getString(R.string.logger_enabled))
  }

  internal fun deleteAllPreviousLogs() {
    loggerWifi.deleteLog(this::onLogDeleted)
    loggerData.deleteLog(this::onLogDeleted)
    loggerBluetooth.deleteLog(this::onLogDeleted)
    loggerSync.deleteLog(this::onLogDeleted)
    loggerAirplane.deleteLog(this::onLogDeleted)
    loggerDoze.deleteLog(this::onLogDeleted)
    loggerManager.deleteLog(this::onLogDeleted)
    loggerTrigger.deleteLog(this::onLogDeleted)
  }

  override fun onDestroyView() {
    super.onDestroyView()
    loggingEnabled.onPreferenceChangeListener = null
  }

  override fun onStart() {
    super.onStart()
    loggerWifi.retrieveLogContents(this::onPrepareLogContentRetrieval, this::onLogContentRetrieved,
        this::onAllLogContentsRetrieved)
    loggerData.retrieveLogContents(this::onPrepareLogContentRetrieval, this::onLogContentRetrieved,
        this::onAllLogContentsRetrieved)
    loggerBluetooth.retrieveLogContents(this::onPrepareLogContentRetrieval,
        this::onLogContentRetrieved, this::onAllLogContentsRetrieved)
    loggerSync.retrieveLogContents(this::onPrepareLogContentRetrieval, this::onLogContentRetrieved,
        this::onAllLogContentsRetrieved)
    loggerAirplane.retrieveLogContents(this::onPrepareLogContentRetrieval,
        this::onLogContentRetrieved, this::onAllLogContentsRetrieved)
    loggerDoze.retrieveLogContents(this::onPrepareLogContentRetrieval, this::onLogContentRetrieved,
        this::onAllLogContentsRetrieved)
    loggerManager.retrieveLogContents(this::onPrepareLogContentRetrieval,
        this::onLogContentRetrieved, this::onAllLogContentsRetrieved)
    loggerTrigger.retrieveLogContents(this::onPrepareLogContentRetrieval,
        this::onLogContentRetrieved, this::onAllLogContentsRetrieved)

    loggerPrefrencePresenter.clickEvent(loggingEnabled, {
      deleteAllPreviousLogs()
    })
  }

  override fun onStop() {
    super.onStop()
    loggerWifi.stop()
    loggerData.stop()
    loggerBluetooth.stop()
    loggerSync.stop()
    loggerAirplane.stop()
    loggerDoze.stop()
    loggerManager.stop()
    loggerTrigger.stop()

    loggerPrefrencePresenter.stop()
  }

  override fun onDestroy() {
    super.onDestroy()
    loggerWifi.destroy()
    loggerData.destroy()
    loggerBluetooth.destroy()
    loggerSync.destroy()
    loggerAirplane.destroy()
    loggerDoze.destroy()
    loggerManager.destroy()
    loggerTrigger.destroy()


    loggerPrefrencePresenter.destroy()
  }

  private val loggerDialog: LoggerDialog
    @CheckResult get() {
      val loggerDialog = parentFragment
      if (loggerDialog is LoggerDialog) {
        return loggerDialog
      } else {
        throw RuntimeException("Parent is not LoggerDialog")
      }
    }

  fun onPrepareLogContentRetrieval() {
    loggerDialog.onPrepareLogContentRetrieval()
  }

  fun onLogContentRetrieved(logLine: String) {
    loggerDialog.onLogContentRetrieved(logLine)
  }

  fun onAllLogContentsRetrieved() {
    loggerDialog.onAllLogContentsRetrieved()
  }

  fun onLogDeleted(logId: String) {
    loggerDialog.onLogDeleted(logId)
  }

  companion object {
    const val TAG = "LoggerPreferenceFragment"
  }
}
